src/upsample.jl

```diff
@@ -64,235 +64,175 @@ function ChainRulesCore.rrule(::typeof(upsample_nearest), x::AbstractArray, s::T
-    weight .= 0
-    for i in 1:n_chan
-        weight[:,:,i,i] .= kern
-    end
-    # weight = cat(fill(kern, n_chan)..., dims=(3,4)) # slow
-    dx = conv(Δ, weight, pad=pad, stride=stride)
-
-    # Still have to fix edge effects due to zero-padding of convolution,
-    # TODO: Could be circumvented by having padding that just extrapolates the value at the first/last index
-    # nextras = tuple((Int.(floor(factor//2)) for factor in k)...)
-    nextras = (floor(Int, k[1]//2), floor(Int, k[2]//2))
-
-    # First dimension edge-effect correction
-    if nextras[1] > 0
-        kern1 = kern[1:nextras[1],:]
-        pad1 = (0, pad[2])
-        stride1 = (1, stride[2])
-        weight1 = similar(Δ, eltype(Δ), (size(kern1)..., n_chan, n_chan))
-        weight1 .= 0
-        for i in 1:n_chan
-            weight1[:,:,i,i] .= kern1
-        end
-        # weight1 = cat(fill(kern1, n_chan)..., dims=(3,4)) # slow
-        dx[[1],:,:,:] .+= conv(Δ[1:nextras[1],:,:,:], weight1, pad=pad1, stride=stride1)
-        weight1 .= weight1[end:-1:1,:,:,:]
-        dx[[end],:,:,:] .+= conv(Δ[end-nextras[1]+1:end,:,:,:], weight1, pad=pad1, stride=stride1)
-    
-        ## Conv with views is not dispatched to CUDA.conv
-        # dx[[1],:,:,:] .+= conv(@view(Δ[1:nextras[1],:,:,:]), weight1, pad=pad1, stride=stride1)
-        # weight1 .= @view(weight1[end:-1:1,:,:,:])
-        # dx[[end],:,:,:] .+= conv(@view(Δ[end-nextras[1]+1:end,:,:,:]), weight1, pad=pad1, stride=stride1)
-    end
-
-    # Second dimension edge-effect correction
-    if nextras[2] > 0
-        kern2 = kern[:,1:nextras[2]]
-        pad2 = (pad[1], 0)
-        stride2 = (stride[1], 1)
-        weight2 = similar(Δ, eltype(Δ), (size(kern2)..., n_chan, n_chan))
-        weight2 .= 0
-        for i in 1:n_chan
-            weight2[:,:,i,i] .= kern2
-        end
-        # weight2 = cat(fill(kern2, n_chan)..., dims=(3,4)) # slow
-    
-        yy = conv(Δ[:,1:nextras[2],:,:], weight2, pad=pad2, stride=stride2)
-        dx[:,[1],:,:] .+= conv(Δ[:,1:nextras[2],:,:], weight2, pad=pad2, stride=stride2)
-        weight2 .= weight2[:,end:-1:1,:,:]
-        dx[:,[end],:,:] .+= conv(Δ[:,end-nextras[2]+1:end,:,:], weight2, pad=pad2, stride=stride2)
-
-        ## Conv with views is not dispatched to CUDA.conv
-        # yy = conv(@view(Δ[:,1:nextras[2],:,:]), weight2, pad=pad2, stride=stride2)
-        # dx[:,[1],:,:] .+= conv(@view(Δ[:,1:nextras[2],:,:]), weight2, pad=pad2, stride=stride2)
-        # weight2 .= @view(weight2[:,end:-1:1,:,:])
-        # dx[:,[end],:,:] .+= conv(@view(Δ[:,end-nextras[2]+1:end,:,:]), weight2, pad=pad2, stride=stride2)
+function ∇upsample_bilinear(Δ::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); size::Union{Nothing,NTuple{2,Integer}}=nothing) where T
+    w,h,c,n = Base.size(Δ)
+    if scale != (1,1) && size !== nothing
+        error("Please provide either scale or size, not both. Got scale=$scale and size=$size.")
     end
```

Comment:
And here, perhaps simplest would be never to accept `scale` at all here? It's more of an internal function, only called within the gradient rule, and at that point the desired size is known exactly. No need to think hard about whether rounding errors up & down with non-integer scale are going to work out correctly.

Reply:
Yes. But @CarloLucibello  implemented having the same call signature for the earlier upsampling methods, forwards and backwards and so did I in consequence. But in principle with the newer implementation, (out)size is enough. Feel free to remove it.

Reply:
OK, I can't make one suggestion for the whole function, but step 1:
```suggestion
function ∇upsample_bilinear(Δ::AbstractArray{T,4}; size::Union{Nothing,NTuple{2,Integer}}) where T
```

Reply:
AAhh sorry man, maybe I have wasted your time here. I think it might be better to leave it for now. Otherwise you also have to fix the tests. And size would need to be positional. And the rrule needs to be changed. My heady is already dizzy.

Reply:
I think the RRule is OK, it always calls with one keyword. It's OK to have a mandatory keyword (less confusing than switching to positional I think).

Reply:
We could have dispatch rules too,

Reply:
Dispatch in which sense? I dont know what you mean here.

Reply:
@CarloLucibello I would like to hear your opinion on the gradient signature, thinking of the discussion in https://github.com/FluxML/NNlib.jl/pull/262